File: NeuralNet/ActivationFunctions/HyperbolicTangensFunction.cs
using System;

namespace NeuralNet {
	[Serializable]
	public sealed class HyperbolicTangensFunction : IActivationFunction {
		private readonly float _alpha;
		private readonly float _betta;
		private readonly float _derivativeFactor;

		public HyperbolicTangensFunction(float alpha, float betta) {
			_alpha = alpha;
			_betta = betta;
			_derivativeFactor = _betta/_alpha;
		}

		public float Calculate(float x) {
			return (_alpha*(float)Math.Tanh(_betta*x));
		}

		public float CalculateFirstDerivative(float x) {
			var tmp = (float)Math.Tanh(_betta*x);
			return _alpha*_betta*(1.0f - tmp*tmp);
		}

		public float CalculateFirstDerivative(float[] state, int index) {
			var neuronState = state[index];
			return _derivativeFactor*(_alpha - neuronState)*(_alpha + neuronState);
		}

		public void CalculateFirstDerivative(float[] target, float[] factors, float[] state) {
			for (var i = 0; i < state.Length; i++) {
				var neuronState = state[i];
				target[i] = factors[i]*_derivativeFactor*(_alpha - neuronState)*(_alpha + neuronState);
			}
		}

		public void CalculateFirstDerivative(float[] target, float[] state) {
			for (var i = 0; i < state.Length; i++) {
				var neuronState = state[i];
				target[i] *= _derivativeFactor*(_alpha - neuronState)*(_alpha + neuronState);
			}
		}

		public float GetMaxDerivativeZone(float maxValuePercent) {
			var value = 1.0/Math.Sqrt(maxValuePercent);
			return (float) Math.Log(value + Math.Sqrt(value*value - 1.0))/_betta;
		}

		public float CalculateInvers(float y) {
			return (float)Math.Log((_alpha + y)/(_alpha - y), (float)Math.E)/(2.0f*_betta);
		}

		public float Alpha {
			get { return _alpha; }
		}

		public float Betta {
			get { return _betta; }
		}
	}
}

File: NeuralNetNative/HyperbolicTangensFunction.cpp
#define NEURALNETNATIVEAPI
#include <mathimf.h>
#include "HyperbolicTangensFunction.h"
#include <tbb\tbb.h>
#include <tbb\task_scheduler_init.h>
#include <tbb\parallel_for.h>
#include <tbb\blocked_range.h>
#include "GrainSizeForParallel.h"

using namespace tbb;

namespace NeuralNetNative {
	HyperbolicTangensFunction::HyperbolicTangensFunction(float alpha, float betta) {
		_alpha = alpha;
		_betta = betta;
		_derivativeFactor = _betta/_alpha;
	}

	float HyperbolicTangensFunction::Calculate(float x) {
		return (_alpha*tanhf(_betta*x));
	}

	float HyperbolicTangensFunction::CalculateFirstDerivative(float x) {
		float tmp = tanhf(_betta*x);
		return _alpha*_betta*(1.0f - tmp*tmp);
	}

	float HyperbolicTangensFunction::CalculateFirstDerivative(const float *state, int index, int stateLength) {
		return _derivativeFactor*(_alpha - state[index])*(_alpha + state[index]);
	}

	void HyperbolicTangensFunction::CalculateFirstDerivative(float* target, const float* factors, const float* state, int stateLength) {
		parallel_for(blocked_range<size_t>(0, stateLength, HyperbolicTangensFunctionGrainSize),
		[=](const blocked_range<size_t>& r)
		{
			for (int i = r.begin(); i < r.end(); i++) {
				target[i] = factors[i]*_derivativeFactor*(_alpha - state[i])*(_alpha + state[i]);
			}
		});
	}

	void HyperbolicTangensFunction::CalculateFirstDerivative(float* target, const float* state, int stateLength) {
		parallel_for(blocked_range<size_t>(0, stateLength, HyperbolicTangensFunctionGrainSize),
		[=](const blocked_range<size_t>& r)
		{
			for (int i = r.begin(); i < r.end(); i++) {
				target[i] *= _derivativeFactor*(_alpha - state[i])*(_alpha + state[i]);
			}
		});
	}

	float HyperbolicTangensFunction::CalculateInvers(float y) {
		return atanhf(y/_alpha)/_betta;
	}
}